packages/core/mx/contracts/staking/src/lib.rs

```diff
@@ -0,0 +1,336 @@
+    fn unstake(&self, tokens: BigUint) {
+        let caller = self.blockchain().get_caller();
+        require!(!self.stakes(&caller).is_empty(), "Caller is not a staker");
+
+        let mut staker = self.stakes(&caller).get();
+        require!(staker.token_staked > 0, "Must have tokens staked");
+        require!(tokens > 0, "Unstake amount must be greater than 0");
+        require!(staker.tokens_available() >= tokens, "Insufficient amount to unstake");
+
+        let new_stake = staker.tokens_secure_stake() - &tokens;
+        require!(new_stake == 0 || new_stake >= self.minimum_stake().get(), "Total stake is below the minimum threshold");
+
+        let block_number = self.blockchain().get_block_nonce();
+        let tokens_to_withdraw = staker.tokens_withdrawable(block_number);
+        if tokens_to_withdraw > 0 {
+            self.withdraw(&mut staker, &caller);
+        }
+
+        staker.lock_tokens(&tokens, self.lock_period().get(), block_number);
+        self.stakes(&caller).set(&staker);
+
+        self.stake_locked_event(&caller, &staker.tokens_locked_until, &tokens);
+    }
+
+    #[endpoint(withdraw)]
+    fn withdraw_endpoint(&self) {
+        let caller = self.blockchain().get_caller();
+        require!(!self.stakes(&caller).is_empty(), "Caller is not a staker");
+
+        let mut staker = self.stakes(&caller).get();
+        self.withdraw(&mut staker, &caller);
+
+        if staker.is_empty() {
+            let mut stakers = self.stakers().get();
+            for (i, staker) in stakers.iter().enumerate() {
+                if *staker == caller {
+                    stakers.remove(i);
+                    self.stakers().set(&stakers);
+                    self.stakes(&caller).clear();
+                    return
+                }
+            }
+        }
+        self.stakes(&caller).set(staker);
+    }
+
+    #[only_owner]
+    #[endpoint(slash)]
+    fn slash(
+        &self,
+        slasher: ManagedAddress,
+        staker_address: ManagedAddress,
+        escrow_address: ManagedAddress,
+        tokens: BigUint,
+    ) {
+        require!(!self.stakes(&staker_address).is_empty(), "Invalid address for staker");
+        let mut staker = self.stakes(&staker_address).get();
+
+        require!(!self.allocations(&escrow_address).is_empty(), "Invalid address for escrow");
+        let mut allocation = self.allocations(&escrow_address).get();
+
+        require!(allocation.tokens > 0, "Allocation must have tokens");
+        require!(allocation.tokens > tokens, "Allocation must have more tokens than the amount to slash");
+        require!(staker.tokens_allocated >= tokens, "Staker must have more tokens than the amount to slash");
+        require!(staker.token_staked >= tokens, "Staker must have more tokens than the amount to slash");
+
+        staker.unallocate(&tokens);
+        staker.release(&tokens);
+        allocation.tokens = allocation.tokens - &tokens;
+
+        self.stakes(&staker_address).set(&staker);
```

Comment:
Here as well, you could use `set()` on the previously saved mapper.

Reply:
modfied to use the `stakes_handler`